[quant][graphmode] Remove observer module after insert_quant_dequant

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28985 [quant][graphmode] Remove observer module after insert_quant_dequant**

Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

jerryzh168 committed:
[quant][graphmode] Remove observer module after insert_quant_dequant

Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -459,16 +459,16 @@ class QuantizeHelper {
     const std::string& observer_name) {
   // remove observer_module
-  observer_modules_to_remove_.push_back(observer_name);
+  module_.module_object()->unsafeRemoveAttr(observer_name);
```
Every `unsafeRemove` has a O(N) complexity where N is the number of slots after the one we're removing. In our case it will be a number of observers. We will do that for each observer, so the total complexity of the algorithm would be O(N^2). Would it be possible to change it to be linear (e.g. delete all of them starting from the last in the very end)?

jerryzh168 replied:
Sure

jerryzh168 replied:
it will complicate the code a bit tho

ZolotukhinM reviewed (approved):
Overall looks good, please find a couple of comments inline.

jerryzh168 committed:
Update on "[quant][graphmode] Remove observer module after insert_quant_dequant"


Summary:
Remove the observer module in the quantized model

Test Plan:
python test/test_jit.py 'TestJit.test_insert_quant_dequant'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253777](https://our.internmc.facebook.com/intern/diff/D18253777)

[ghstack-poisoned]
